File: 99-recover-binary-search-tree/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def recoverTree(self, root):
        """
        :type root: TreeNode
        :rtype: void Do not return anything, modify root in-place instead.
        """
        values = []
        self.inorder(root, values)
        vals = [n.val for n in sorted(values, key=lambda v: v.val)]
        for i in range(len(values)):
            values[i].val = vals[i]

    def inorder(self, root, values):
        if root.left:
            self.inorder(root.left, values)
        values.append(root)
        if root.right:
            self.inorder(root.right, values)
        
        

File: 99-recover-binary-search-tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void recoverTree(TreeNode* root) {
        vector<TreeNode*> vs;
        vector<int> vals;
        inorder(root, vs, vals);
        sort(vals.begin(), vals.end());
        for (int i = 0; i < vs.size(); i++) vs[i]->val = vals[i];
    }
    
    void inorder(TreeNode* root, vector<TreeNode*> & vs, vector<int> & vals) {
        if (root->left) inorder(root->left, vs, vals);
        vs.push_back(root);
        vals.push_back(root->val);
        if (root->right) inorder(root->right, vs, vals);
    }
};